File: csharp/extractor/Semmle.Extraction.CSharp/CachedEntityFactory.cs
using Microsoft.CodeAnalysis;

namespace Semmle.Extraction.CSharp
{
    /// <summary>
    /// A factory for creating cached entities.
    /// </summary>
    internal abstract class CachedEntityFactory<TInit, TEntity>
        : Extraction.CachedEntityFactory<TInit, TEntity> where TEntity : CachedEntity
    {
        /// <summary>
        /// Initializes the entity, but does not generate any trap code.
        /// </summary>
        public sealed override TEntity Create(Extraction.Context cx, TInit init)
        {
            return Create((Context)cx, init);
        }

        public override bool IsShared(TInit init)
        {
            return
                base.IsShared(init) &&
                !(init is ITypeSymbol type && type.ContainsAnonymousType()) &&
                !(init is IMethodSymbol method && method.ContainsAnonymousType()) &&
                !(init is ISymbol symbol && symbol.ContainingType?.ContainsAnonymousType() == true);
        }

        public abstract TEntity Create(Context cx, TInit init);
    }
}

File: java/kotlin-extractor2/src/main/kotlin/entities/CachedEntityFactory.kt
package com.github.codeql.entities

import com.github.codeql.AnyDbType
import org.jetbrains.kotlin.analysis.api.KaSession
import org.jetbrains.kotlin.analysis.api.symbols.KaSymbol

abstract class CachedEntityFactory<TSymbol : KaSymbol, TEntity : Entity<TSymbol, AnyDbType>> {
    context(KaSession)
    fun create(context: Context, symbol: TSymbol): TEntity {
        return context.createEntity(this, symbol)
    }

    context(KaSession)
    abstract fun createEntity(context: Context, symbol: TSymbol): TEntity
}